iotdb-core/datanode/src/main/java/org/apache/iotdb/db/pipe/task/subtask/connector/PipeRealtimePriorityBlockingQueue.java

```diff
@@ -71,30 +78,53 @@ public boolean put(final Event event) {
 
   @Override
   public Event directPoll() {
-    Event event = super.directPoll();
+    Event event = null;
+    if (eventCount.get() >= realTimeEventConsumeThreshold) {
+      event = tsfileInsertEventDeque.pollFirst();
+      eventCount.set(0);
+    }
     if (Objects.isNull(event)) {
-      event = tsfileInsertEventDeque.pollLast();
+      event = super.directPoll();
+      if (Objects.isNull(event)) {
+        event = tsfileInsertEventDeque.pollLast();
+      }
+      if (event != null) {
+        eventCount.incrementAndGet();
+      }
     }
+
     return event;
   }
 
   /**
-   * Try to poll the freshest insertion event from the queue. First, try to poll the first offered
-   * non-TsFileInsertionEvent. If no such event is available, poll the last offered
-   * TsFileInsertionEvent. If no event is available, block until an event is available.
+   * When the number of polls exceeds the realtimeConsumeThreshold, the {@link TsFileInsertionEvent}
+   * of the earliest write to the queue is returned. if the realtimeConsumeThreshold is not reached
+   * then an attempt is made to poll the queue for the latest insertion {@link Event}. First, it
+   * tries to poll the first provided If there is no such {@link Event}, poll the last supplied
+   * {@link TsFileInsertionEvent}. If no {@link Event} is available, it blocks until a {@link Event}
+   * is available.
    *
-   * @return the freshest insertion event. can be null if no event is available.
+   * @return the freshest insertion {@link Event}. can be {@code null} if no event is available.
    */
   @Override
   public Event waitedPoll() {
     Event event = null;
-
-    if (!super.isEmpty()) {
+    if (eventCount.get() >= realTimeEventConsumeThreshold) {
+      event = tsfileInsertEventDeque.pollFirst();
+      if (event != null) {
+        eventCount.set(0);
+      }
+    }
+    if (event == null && !super.isEmpty()) {
```

Comment:
If super.isEmpty(), the tsFile in the deque will not polled here. It may be polled in the "block" code before waiting for a period of time if super.isEmpty() is always true(usually 5s). Adjust the code logic to achieve consistency with the previous logic.

Reply:
May delete the "isEmpty" judgment directly...